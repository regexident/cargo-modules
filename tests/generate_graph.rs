#[macro_use]
mod util;

mod colors {
    mod plain {
        test_cmd!(
            args: "generate graph",
            success: true,
            color_mode: ColorMode::Plain,
            project: smoke
        );
    }

    mod ansi {
        test_cmd!(
            args: "generate graph",
            success: true,
            color_mode: ColorMode::Ansi,
            project: smoke
        );
    }

    mod truecolor {
        test_cmd!(
            args: "generate graph",
            success: true,
            color_mode: ColorMode::TrueColor,
            project: smoke
        );
    }
}

mod default {
    mod pass {
        test_cmds!(
            args: "generate graph",
            success: true,
            color_mode: ColorMode::Plain,
            projects: [
                package_bin_target,
                package_lib_target,
                virtual_workspace_single_package_lib_target,
                virtual_workspace_single_package_bin_target,
                workspace_single_package_lib_target,
                workspace_single_package_bin_target,
            ]
        );
    }

    mod fail {
        test_cmds!(
            args: "generate graph",
            success: false,
            color_mode: ColorMode::Plain,
            projects: [
                package_multi_target,
                virtual_workspace_single_package_multi_target,
                workspace_single_package_multi_target,
            ]
        );
    }
}

mod negative_args {
    use clap::Parser;

    use cargo_modules::{commands::Command::Generate, generate::Command, options::App};

    fn args_for(command: &str, arg: &str, default: bool) -> Vec<(Vec<String>, bool)> {
        let args_prefix = vec![
            "modules".to_owned(),
            "generate".to_owned(),
            command.to_owned(),
        ];

        let pos_arg = format!("--{arg}");
        let neg_arg = format!("--no-{arg}");

        let arg_suffixes = vec![
            (vec![], default),
            (vec![pos_arg.clone()], true),
            (vec![neg_arg.clone()], false),
            (vec![pos_arg.clone(), neg_arg.clone()], false),
            (vec![neg_arg.clone(), pos_arg.clone()], true),
        ];

        arg_suffixes
            .into_iter()
            .map(|(args_suffix, expected)| {
                let mut args = args_prefix.clone();
                args.extend(args_suffix);
                (args, expected)
            })
            .collect()
    }

    mod project {
        use super::*;

        #[test]
        fn cfg_test() {
            for command in ["tree", "graph"] {
                for (args, expected) in args_for(command, "cfg-test", false) {
                    let app = App::parse_from(&args);

                    let Generate { cmd: Command::Graph(cmd) } = app.command else { continue; };

                    assert_eq!(cmd.project.cfg_test, expected, "{:?}", args);
                }
            }
        }

        #[test]
        fn sysroot() {
            for command in ["tree", "graph"] {
                for (args, expected) in args_for(command, "sysroot", false) {
                    let app = App::parse_from(&args);

                    let Generate { cmd: Command::Graph(cmd) } = app.command else { continue; };

                    assert_eq!(cmd.project.sysroot, expected, "{:?}", args);
                }
            }
        }
    }

    mod graph {
        use super::*;

        #[test]
        fn fns() {
            for command in ["tree", "graph"] {
                for (args, expected) in args_for(command, "fns", false) {
                    let app = App::parse_from(&args);

                    let Generate { cmd: Command::Graph(cmd) } = app.command else { continue; };

                    assert_eq!(cmd.graph.fns, expected, "{:?}", args);
                }
            }
        }

        #[test]
        fn orphans() {
            for command in ["tree", "graph"] {
                for (args, expected) in args_for(command, "orphans", false) {
                    let app = App::parse_from(&args);

                    let Generate { cmd: Command::Graph(cmd) } = app.command else { continue; };

                    assert_eq!(cmd.graph.orphans, expected, "{:?}", args);
                }
            }
        }

        #[test]
        fn tests() {
            for command in ["tree", "graph"] {
                for (args, expected) in args_for(command, "tests", false) {
                    let app = App::parse_from(&args);

                    let Generate { cmd: Command::Graph(cmd) } = app.command else { continue; };

                    assert_eq!(cmd.graph.tests, expected, "{:?}", args);
                }
            }
        }

        #[test]
        fn types() {
            for command in ["tree", "graph"] {
                for (args, expected) in args_for(command, "types", false) {
                    let app = App::parse_from(&args);

                    let Generate { cmd: Command::Graph(cmd) } = app.command else { continue; };

                    assert_eq!(cmd.graph.types, expected, "{:?}", args);
                }
            }
        }
    }

    mod graph_only {
        use super::*;

        #[test]
        fn modules() {
            for (args, expected) in args_for("graph", "modules", true) {
                let app = App::parse_from(&args);

                let Generate { cmd: Command::Graph(cmd) } = app.command else { continue; };

                assert_eq!(cmd.modules, expected, "{:?}", args);
            }
        }

        #[test]
        fn uses() {
            for (args, expected) in args_for("graph", "uses", false) {
                let app = App::parse_from(&args);

                let Generate { cmd: Command::Graph(cmd) } = app.command else { continue; };

                assert_eq!(cmd.uses, expected, "{:?}", args);
            }
        }

        #[test]
        fn externs() {
            for (args, expected) in args_for("graph", "externs", false) {
                let app = App::parse_from(&args);

                let Generate { cmd: Command::Graph(cmd) } = app.command else { continue; };

                assert_eq!(cmd.externs, expected, "{:?}", args);
            }
        }
    }
}

mod lib {
    mod pass {
        test_cmds!(
            args: "generate graph \
                    --lib",
            success: true,
            color_mode: ColorMode::Plain,
            projects: [
                package_lib_target,
                package_multi_target,
                virtual_workspace_single_package_lib_target,
                virtual_workspace_single_package_multi_target,
                workspace_single_package_lib_target,
                workspace_single_package_multi_target,
            ]
        );
    }

    mod fail {
        test_cmds!(
            args: "generate graph \
                    --lib", // does not exist
            success: false,
            color_mode: ColorMode::Plain,
            projects: [
                package_bin_target,
                virtual_workspace_single_package_bin_target,
                workspace_single_package_bin_target,
            ]
        );
    }
}

mod bin {
    mod pass {
        test_cmds!(
            args: "generate graph \
                    --bin package_bin_target",
            success: true,
            color_mode: ColorMode::Plain,
            projects: [
                package_bin_target,
                virtual_workspace_single_package_bin_target,
                workspace_single_package_bin_target,
            ]
        );

        test_cmds!(
            args: "generate graph \
                    --bin package_multi_target",
            success: true,
            color_mode: ColorMode::Plain,
            projects: [
                package_multi_target,
                virtual_workspace_single_package_multi_target,
                workspace_single_package_multi_target,
            ]
        );
    }

    mod fail {
        test_cmds!(
            args: "generate graph \
                    --bin foobar", // does not exist
            success: false,
            color_mode: ColorMode::Plain,
            projects: [
                package_lib_target,
                package_multi_target,
                virtual_workspace_multi_package,
                virtual_workspace_single_package_lib_target,
                virtual_workspace_single_package_multi_target,
                workspace_multi_package,
                workspace_single_package_lib_target,
                workspace_single_package_multi_target,
            ]
        );
    }
}

mod package {
    mod pass {
        test_cmds!(
            args: "generate graph \
                    --package package_lib_target",
            success: true,
            color_mode: ColorMode::Plain,
            projects: [
                package_lib_target,
                virtual_workspace_single_package_lib_target,
                workspace_single_package_lib_target,
            ]
        );

        test_cmds!(
            args: "generate graph \
                    --package package_bin_target",
            success: true,
            color_mode: ColorMode::Plain,
            projects: [
                package_bin_target,
                virtual_workspace_single_package_bin_target,
                workspace_single_package_bin_target,
            ]
        );
    }

    mod fail {
        test_cmds!(
            args: "generate graph \
                    --package foobar",
            success: false,
            color_mode: ColorMode::Plain,
            projects: [
                package_bin_target,
                package_lib_target,
                package_multi_target,
                virtual_workspace_multi_package,
                virtual_workspace_single_package_bin_target,
                virtual_workspace_single_package_lib_target,
                virtual_workspace_single_package_multi_target,
                workspace_multi_package,
                workspace_single_package_bin_target,
                workspace_single_package_lib_target,
                workspace_single_package_multi_target,
            ]
        );
    }
}

mod package_lib {
    mod pass {
        test_cmds!(
            args: "generate graph \
                    --package package_lib_target \
                    --lib",
            success: true,
            color_mode: ColorMode::Plain,
            projects: [
                package_lib_target,
                workspace_single_package_lib_target,
                virtual_workspace_single_package_lib_target,
            ]
        );

        test_cmds!(
            args: "generate graph \
                    --package package_multi_target \
                    --lib",
            success: true,
            color_mode: ColorMode::Plain,
            projects: [
                package_multi_target,
                workspace_single_package_multi_target,
                virtual_workspace_single_package_multi_target,
            ]
        );
    }

    mod fail {
        test_cmds!(
            args: "generate graph \
                    --package package_bin_target \
                    --lib", // does not exist
            success: false,
            color_mode: ColorMode::Plain,
            projects: [
                package_bin_target,
                workspace_single_package_bin_target,
                virtual_workspace_single_package_bin_target,
            ]
        );
    }
}

mod package_bin {
    mod pass {
        test_cmds!(
            args: "generate graph \
                    --package package_bin_target \
                    --bin package_bin_target",
            success: true,
            color_mode: ColorMode::Plain,
            projects: [
                package_bin_target,
                workspace_single_package_bin_target,
                virtual_workspace_single_package_bin_target,
            ]
        );

        test_cmds!(
            args: "generate graph \
                    --package package_multi_target \
                    --bin package_multi_target",
            success: true,
            color_mode: ColorMode::Plain,
            projects: [
                package_multi_target,
                workspace_single_package_multi_target,
                virtual_workspace_single_package_multi_target,
            ]
        );
    }

    mod fail {
        test_cmds!(
            args: "generate graph \
                    --package workspace-package \
                    --bin foobar", // does not exist
            success: false,
            color_mode: ColorMode::Plain,
            projects: [
                package_bin_target,
                package_lib_target,
                package_multi_target,
                virtual_workspace_multi_package,
                virtual_workspace_single_package_bin_target,
                virtual_workspace_single_package_lib_target,
                virtual_workspace_single_package_multi_target,
                workspace_multi_package,
                workspace_single_package_bin_target,
                workspace_single_package_lib_target,
                workspace_single_package_multi_target,
            ]
        );
    }
}

mod orphans {
    test_cmd!(
        args: "generate graph \
                --orphans",
        success: true,
        color_mode: ColorMode::Plain,
        project: smoke
    );
}

mod tests {
    test_cmd!(
        args: "generate graph \
                --tests",
        success: true,
        color_mode: ColorMode::Plain,
        project: smoke
    );
}

mod types {
    test_cmd!(
        args: "generate graph \
                --types",
        success: true,
        color_mode: ColorMode::Plain,
        project: smoke
    );
}

mod types_with_no_modules {
    test_cmd!(
        args: "generate graph \
                --types \
                --no-modules",
        success: true,
        color_mode: ColorMode::Plain,
        project: smoke
    );
}

mod traits {
    test_cmd!(
        args: "generate graph \
                --traits",
        success: true,
        color_mode: ColorMode::Plain,
        project: smoke
    );
}

mod fns {
    test_cmd!(
        args: "generate graph \
                --fns",
        success: true,
        color_mode: ColorMode::Plain,
        project: smoke
    );
}

mod no_modules {
    test_cmd!(
        args: "generate graph \
                --no-modules",
        success: true,
        color_mode: ColorMode::Plain,
        project: smoke
    );
}

mod uses {
    test_cmd!(
        args: "generate graph \
                --uses",
        success: true,
        color_mode: ColorMode::Plain,
        project: smoke
    );
}

mod externs {
    test_cmd!(
        args: "generate graph \
                --externs",
        success: false,
        color_mode: ColorMode::Plain,
        project: smoke
    );
}

mod uses_with_externs {
    test_cmd!(
        args: "generate graph \
                --uses \
                --externs",
        success: true,
        color_mode: ColorMode::Plain,
        project: smoke
    );
}

mod uses_with_externs_with_sysroot {
    test_cmd!(
        attrs: [
            // `sysroot` is expensive, so only run on release builds:
            #[ignore]
        ],
        args: "generate graph \
                --uses \
                --externs \
                --sysroot",
        success: true,
        color_mode: ColorMode::Plain,
        project: smoke
    );
}

mod focus_on {
    mod simple_path {
        test_cmd!(
            args: "generate graph \
                    --uses \
                    --focus-on \"smoke::visibility::dummy\"",
            success: true,
            color_mode: ColorMode::Plain,
            project: smoke
        );
    }
    mod glob_path {
        test_cmd!(
            args: "generate graph \
                    --uses \
                    --focus-on \"smoke::visibility::*\"",
            success: true,
            color_mode: ColorMode::Plain,
            project: smoke
        );
    }
    mod self_path {
        test_cmd!(
            args: "generate graph \
                    --uses \
                    --focus-on \"smoke::visibility::dummy::{self}\"",
            success: true,
            color_mode: ColorMode::Plain,
            project: smoke
        );
    }
    mod tree {
        test_cmd!(
            args: "generate graph \
                    --uses \
                    --focus-on \"smoke::visibility::{dummy, hierarchy}\"",
            success: true,
            color_mode: ColorMode::Plain,
            project: smoke
        );
    }
}

mod max_depth {
    mod depth_0 {
        test_cmd!(
            args: "generate graph \
                    --uses \
                    --max-depth 0",
            success: true,
            color_mode: ColorMode::Plain,
            project: smoke
        );
    }

    mod depth_1 {
        test_cmd!(
            args: "generate graph \
                    --uses \
                    --max-depth 1",
            success: true,
            color_mode: ColorMode::Plain,
            project: smoke
        );
    }

    mod depth_2 {
        test_cmd!(
            args: "generate graph \
                    --uses \
                    --max-depth 2",
            success: true,
            color_mode: ColorMode::Plain,
            project: smoke
        );
    }
}

mod github_issue_79 {
    test_cmd!(
        args: "generate graph \
                --uses",
        success: true,
        color_mode: ColorMode::Plain,
        project: github_issue_79
    );
}

mod github_issue_80 {
    mod tests {
        test_cmd!(
            args: "generate graph \
                    --uses \
                    --types \
                    --tests",
            success: true,
            color_mode: ColorMode::Plain,
            project: github_issue_80
        );
    }

    mod without_tests {
        test_cmd!(
            args: "generate graph \
                    --uses \
                    --types",
            success: true,
            color_mode: ColorMode::Plain,
            project: github_issue_80
        );
    }
}

mod github_issue_172 {
    test_cmd!(
        args: "generate graph \
                --types \
                --uses \
                --traits \
                --layout dot",
        success: true,
        color_mode: ColorMode::Plain,
        project: github_issue_172
    );
}
