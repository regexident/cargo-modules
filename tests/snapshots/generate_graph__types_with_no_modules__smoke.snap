---
source: tests/generate_graph.rs
expression: output
---
STDERR:

STDOUT:
digraph {

    graph [
        label="smoke",
        labelloc=t,

        pad=0.4,

        // Consider rendering the graph using a different layout algorithm, such as:
        // [dot, neato, twopi, circo, fdp, sfdp]
        layout=neato,
        overlap=false,
        splines="line",
        rankdir=LR,

        fontname="Helvetica", 
        fontsize="36",
    ];

    node [
        fontname="monospace",
        fontsize="10",
        shape="record",
        style="filled",
    ];

    edge [
        fontname="monospace",
        fontsize="10",
    ];

    "smoke" [label="crate|smoke", fillcolor="#5397c8"]; // "crate" node
    "smoke::hierarchy::lorem::Lorem" [label="pub(self) struct|hierarchy::lorem::Lorem", fillcolor="#db5367"]; // "struct" node
    "smoke::hierarchy::lorem::consectetur::Consectetur" [label="pub(self) struct|hierarchy::lorem::consectetur::Consectetur", fillcolor="#db5367"]; // "struct" node
    "smoke::hierarchy::lorem::consectetur::adipiscing::Adipiscing" [label="pub(self) struct|hierarchy::lorem::consectetur::adipiscing::Adipiscing", fillcolor="#db5367"]; // "struct" node
    "smoke::hierarchy::lorem::consectetur::adipiscing::elit::Elit" [label="pub(self) struct|hierarchy::lorem::consectetur::adipiscing::elit::Elit", fillcolor="#db5367"]; // "struct" node
    "smoke::hierarchy::lorem::dolor::Dolor" [label="pub(self) struct|hierarchy::lorem::dolor::Dolor", fillcolor="#db5367"]; // "struct" node
    "smoke::hierarchy::lorem::dolor::sit::Sit" [label="pub(self) struct|hierarchy::lorem::dolor::sit::Sit", fillcolor="#db5367"]; // "struct" node
    "smoke::hierarchy::lorem::dolor::sit::amet::Amet" [label="pub(self) struct|hierarchy::lorem::dolor::sit::amet::Amet", fillcolor="#db5367"]; // "struct" node
    "smoke::hierarchy::lorem::ipsum::Ipsum" [label="pub(self) struct|hierarchy::lorem::ipsum::Ipsum", fillcolor="#db5367"]; // "struct" node
    "smoke::visibility::dummy::enums::PubCrate" [label="pub(crate) enum|visibility::dummy::enums::PubCrate", fillcolor="#f8c04c"]; // "enum" node
    "smoke::visibility::dummy::enums::PubModule" [label="pub(in crate::visibility) enum|visibility::dummy::enums::PubModule", fillcolor="#fe9454"]; // "enum" node
    "smoke::visibility::dummy::enums::PubPrivate" [label="pub(self) enum|visibility::dummy::enums::PubPrivate", fillcolor="#db5367"]; // "enum" node
    "smoke::visibility::dummy::enums::PubPublic" [label="pub enum|visibility::dummy::enums::PubPublic", fillcolor="#81c169"]; // "enum" node
    "smoke::visibility::dummy::enums::PubSuper" [label="pub(super) enum|visibility::dummy::enums::PubSuper", fillcolor="#fe9454"]; // "enum" node
    "smoke::visibility::dummy::structs::PubCrate" [label="pub(crate) struct|visibility::dummy::structs::PubCrate", fillcolor="#f8c04c"]; // "struct" node
    "smoke::visibility::dummy::structs::PubModule" [label="pub(in crate::visibility) struct|visibility::dummy::structs::PubModule", fillcolor="#fe9454"]; // "struct" node
    "smoke::visibility::dummy::structs::PubPrivate" [label="pub(self) struct|visibility::dummy::structs::PubPrivate", fillcolor="#db5367"]; // "struct" node
    "smoke::visibility::dummy::structs::PubPublic" [label="pub struct|visibility::dummy::structs::PubPublic", fillcolor="#81c169"]; // "struct" node
    "smoke::visibility::dummy::structs::PubSuper" [label="pub(super) struct|visibility::dummy::structs::PubSuper", fillcolor="#fe9454"]; // "struct" node
    "smoke::visibility::dummy::unions::PubCrate" [label="pub(crate) union|visibility::dummy::unions::PubCrate", fillcolor="#f8c04c"]; // "union" node
    "smoke::visibility::dummy::unions::PubModule" [label="pub(in crate::visibility) union|visibility::dummy::unions::PubModule", fillcolor="#fe9454"]; // "union" node
    "smoke::visibility::dummy::unions::PubPrivate" [label="pub(self) union|visibility::dummy::unions::PubPrivate", fillcolor="#db5367"]; // "union" node
    "smoke::visibility::dummy::unions::PubPublic" [label="pub union|visibility::dummy::unions::PubPublic", fillcolor="#81c169"]; // "union" node
    "smoke::visibility::dummy::unions::PubSuper" [label="pub(super) union|visibility::dummy::unions::PubSuper", fillcolor="#fe9454"]; // "union" node

    "smoke" -> "smoke::hierarchy::lorem::Lorem" [label="owns", color="#000000", style="solid"] [constraint=true]; // "owns" edge
    "smoke" -> "smoke::hierarchy::lorem::consectetur::Consectetur" [label="owns", color="#000000", style="solid"] [constraint=true]; // "owns" edge
    "smoke" -> "smoke::hierarchy::lorem::consectetur::adipiscing::Adipiscing" [label="owns", color="#000000", style="solid"] [constraint=true]; // "owns" edge
    "smoke" -> "smoke::hierarchy::lorem::consectetur::adipiscing::elit::Elit" [label="owns", color="#000000", style="solid"] [constraint=true]; // "owns" edge
    "smoke" -> "smoke::hierarchy::lorem::dolor::Dolor" [label="owns", color="#000000", style="solid"] [constraint=true]; // "owns" edge
    "smoke" -> "smoke::hierarchy::lorem::dolor::sit::Sit" [label="owns", color="#000000", style="solid"] [constraint=true]; // "owns" edge
    "smoke" -> "smoke::hierarchy::lorem::dolor::sit::amet::Amet" [label="owns", color="#000000", style="solid"] [constraint=true]; // "owns" edge
    "smoke" -> "smoke::hierarchy::lorem::ipsum::Ipsum" [label="owns", color="#000000", style="solid"] [constraint=true]; // "owns" edge
    "smoke" -> "smoke::visibility::dummy::enums::PubCrate" [label="owns", color="#000000", style="solid"] [constraint=true]; // "owns" edge
    "smoke" -> "smoke::visibility::dummy::enums::PubModule" [label="owns", color="#000000", style="solid"] [constraint=true]; // "owns" edge
    "smoke" -> "smoke::visibility::dummy::enums::PubPrivate" [label="owns", color="#000000", style="solid"] [constraint=true]; // "owns" edge
    "smoke" -> "smoke::visibility::dummy::enums::PubPublic" [label="owns", color="#000000", style="solid"] [constraint=true]; // "owns" edge
    "smoke" -> "smoke::visibility::dummy::enums::PubSuper" [label="owns", color="#000000", style="solid"] [constraint=true]; // "owns" edge
    "smoke" -> "smoke::visibility::dummy::structs::PubCrate" [label="owns", color="#000000", style="solid"] [constraint=true]; // "owns" edge
    "smoke" -> "smoke::visibility::dummy::structs::PubModule" [label="owns", color="#000000", style="solid"] [constraint=true]; // "owns" edge
    "smoke" -> "smoke::visibility::dummy::structs::PubPrivate" [label="owns", color="#000000", style="solid"] [constraint=true]; // "owns" edge
    "smoke" -> "smoke::visibility::dummy::structs::PubPublic" [label="owns", color="#000000", style="solid"] [constraint=true]; // "owns" edge
    "smoke" -> "smoke::visibility::dummy::structs::PubSuper" [label="owns", color="#000000", style="solid"] [constraint=true]; // "owns" edge
    "smoke" -> "smoke::visibility::dummy::unions::PubCrate" [label="owns", color="#000000", style="solid"] [constraint=true]; // "owns" edge
    "smoke" -> "smoke::visibility::dummy::unions::PubModule" [label="owns", color="#000000", style="solid"] [constraint=true]; // "owns" edge
    "smoke" -> "smoke::visibility::dummy::unions::PubPrivate" [label="owns", color="#000000", style="solid"] [constraint=true]; // "owns" edge
    "smoke" -> "smoke::visibility::dummy::unions::PubPublic" [label="owns", color="#000000", style="solid"] [constraint=true]; // "owns" edge
    "smoke" -> "smoke::visibility::dummy::unions::PubSuper" [label="owns", color="#000000", style="solid"] [constraint=true]; // "owns" edge

}

